### ISSUE TITLE
[13.0] l10n_latam_invoice_document: Does not allow to use Odoo's internal sequence

### ISSUE THREAD

**Issue by gustavovalverde** [odoo/odoo#43313]
#### Impacted versions:
13.0
#### Steps to reproduce:
- Create an invoice using `l10n_latam_invoice_document`
- The move `name` is overwritten by the field `l10n_latam_document_number`

#### Current behavior:
This causes the move to have 2 fields with the same information:
1. `l10n_latam_invoice_document`
2. `name`

This also _hard-code_ this behavior for other localizations, not allowing them to use `l10n_latam_document_number` as their fiscal number and `name` as the internal sequence for continuity and audit purposes.

#### Expected behavior:
Allow for better inheritance (or logic) of the following methods:
https://github.com/odoo/odoo/blob/23e2824794b64624c8dbb1b8132d443657781830/addons/l10n_latam_invoice_document/models/account_move.py#L31-L41
Mainly line **39**

Right now we have to overwrite the whole method and do:
`rec.l10n_latam_document_number = rec.l10n_latam_document_number`

https://github.com/odoo/odoo/blob/23e2824794b64624c8dbb1b8132d443657781830/addons/l10n_latam_invoice_document/models/account_move.py#L43-L52
Mainly line **52**

Right now we have to do the following to succeed on using both fields:

```python
    @api.onchange('l10n_latam_document_type_id', 'l10n_latam_document_number')
    def _inverse_l10n_latam_document_number(self):
        moves = self.filtered(lambda m: m.l10n_latam_country_code != 'DO')

        for rec in moves.filtered('l10n_latam_document_type_id'):
            if not rec.l10n_latam_document_number:
                rec.name = '/'
            else:
                l10n_latam_document_number = (rec.l10n_latam_document_type_id
                                              ._format_document_number(
                                                  rec.l10n_latam_document_number)
                                              )
                if rec.l10n_latam_document_number != l10n_latam_document_number:
                    rec.l10n_latam_document_number = l10n_latam_document_number

        super(AccountMove, moves)._inverse_l10n_latam_document_number()
```

Video/Screenshot link (optional):

**Comment by gustavovalverde** [comment]
cc: @jco-odoo @zaoral @jjscarafia

**Comment by gustavovalverde** [comment]
I forgot, that this method also requires this PR, so the sequence is not consumed multiple times by one `post()` https://github.com/odoo/odoo/pull/43195/files

**Note:** It's also required to set `l10n_latam_document_number` as `store=True` (which can be done on the localization that inherits this module; so not a blocking issue there, but it would be preferable to make it stored in _master_)

**Comment by jjscarafia** [comment]
Hi @gustavovalverde thanks for the detailed issue.
This would be a major change on the module or at least on the way the module was thought.
In Argentina and Chile, the two countries that original motivate this module, the idea of replacing the "account.move.name" with the document type, is related to:

1. on every report we want to see document number (invoice number), not an internal journal entry number. The only way to do this is using the name (at least the only way accepted by odoo). (for eg. on follow ups, partner ledger, etc)
2. in Arg and Chile the numbering of journals entries is not per journal but globally. Also, at least in Argentina, it's common to summarize many entries on one monthly entry (so that when you print the "Diary book" it doesn't use lot of pages).
3. so, the legal journal entry numbering is not actually implemented when using this module, considering previous ideas. Most companies don't need/use that (at least SMED). And for those who need it there would be two approaches: a new field with journal entry number or to number journal entries on a report.

Is your case different thant Arg/Chile. Isn't '1' the case for you?

**Comment by gustavovalverde** [comment]
Hi @jjscarafia, thanks for your answers.

Basically; I do agree with statement **1**. We do actually 'need' (for convenience) the _document number_ on reports. So far (up to v12), we've managed to have it were it matters the most,on follow-ups, general ledger, account moves, bank reconciliation and (mainly) on search results, using the **ref** field.

I'm afraid of using `account.move.name` mainly because of two things:

1. A bunch of clients will complain because they don't have a number to **track** the account moves and know (by seeing) if one is missing. Most of them are expecting integrity and continuity on the account move sequence.

2. We use the document number the supplier **gives us**, which we expect won't duplicate on the same supplier...but it will duplicate on different supplier. That means a journal having the same `account.move.name` multiple times. How do you deal with this?

Because of the above, from v10 - v12, we've been using:
- `account.invoice.number` --> Internal Sequence
- `account.invoice.ref`--> Document Number
- `account.invoice.name` --> Description used by the user

On v13, based on your answers, this could work for us:
- `account.move.name` --> Internal Sequence
- `account.invoice.ref`--> Document Number (in addition to the one added by l10m_latam_*)
- `account.invoice.narration` --> Description used by the user

**Comment by gustavovalverde** [comment]
I also remembered that the idea of using the `ref` field was given by Odoo, so we just refactored based on that: https://github.com/odoo/odoo/issues/19722#issuecomment-423982982 and, to be honest, we've preferred it that way since then. 

cc: @jco-odoo

**Comment by gustavovalverde** [comment]
By the way, @jjscarafia. I'm not willing to break how **l10n_latam_invoice_document** is working so far for Chile and Argentina. I'm just trying to find an **stable**, **healthy** and **_odoo-like_** way to allow both/multiple use-cases with the same code.

**Comment by jjscarafia** [comment]
Hi @gustavovalverde 
Yes, we fight a lot for this to be accepted by odoo, our first approach (that is the approach we've [here](https://github.com/ingadhoc/account-financial-tools/blob/13.0/account_document/models/account_move.py#L30) from v8, was to add a new field "document_number", inherit the display_name to choose if if move name or move.document_number, and then modify many odoo reports and use .display_name instead of .name.

That was not accepted and they accept actual approach that is actually ok for us (and simpler).

Regarding your two questions:
1. I understand, for the ones who ask for it we would create a new field, also for argentina only populated with once sequence, no one sequence per journal (of course the sequence can be the same en every journal also).

2. We [modified the uniqueness constraint](https://github.com/odoo/odoo/blob/13.0/addons/l10n_latam_invoice_document/models/account_move.py#L105-L107) and the module adds [it's own](https://github.com/odoo/odoo/blob/13.0/addons/l10n_latam_invoice_document/models/account_move.py#L201-L216) to make what you said. NOTE: I believe there is a bug on how we modify the original, we'll check this.

I've share to you one of the documents we've used for the discussion of document_number vs name field. (there where many more discussions but they where almost a year ago).

If you want to test how it looks like the usage on l10n_ar choose any runbot v13 and install l10n_ar. You will have a demo company "Responsable Inscripto" with many invoices and bills, this is what the user expects in Argentina.

Regarding 

> On v13, based on your answers, this could work for us:
> 
> account.move.name --> Internal Sequence
> account.invoice.ref--> Document Number (in addition to the one added by l10m_latam_*)
> account.invoice.narration --> Description used by the user

We didn't do like that because we really want to see document number everywhere, not the sequence number. What we want to see everywhere is the "docuemnt number" + a "reference" that gives more information about the invoice or payment (actual ref field). That's why we cant'use just ref.

> I'm not willing to break how l10n_latam_invoice_document is working so far for Chile and Argentina. I'm just trying to find an stable, healthy and odoo-like way to allow both/multiple use-cases with the same code.
Great! I'm sure we'll find a way. It would be great if you could also join using this module. 

> So far I'm thinking on refactoring this methods and allow them to check for a context, and based on this context decide where l10n_latam_document_number will be written to (name or ref)
This seams less danger than turning the field into an stored one.

I will write to you by email to schedule a hangout if you agree.

Best regards,
Juan

**Comment by jjscarafia** [comment]
I remember now, we share some ideas on [this old issue](https://github.com/odoo/odoo/issues/14059) regarding document_number

**Comment by gustavovalverde** [comment]
Fixed by: #43850
